File: src/main/java/plugin/javafxtools/service/DevEnvironmentService.java
package plugin.javafxtools.service;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/** Performs an on-demand health check of common local development tools. */
public final class DevEnvironmentService {
    private static final Duration TOOL_TIMEOUT = Duration.ofSeconds(5);
    private static final int MAX_OUTPUT_BYTES = 256 * 1024;
    private static final Charset NATIVE_CHARSET = Charset.forName(
            System.getProperty("native.encoding", "UTF-8"));
    private static final List<ToolDefinition> TOOLS = List.of(
            new ToolDefinition("jdk", "JDK 编译器", "javac", List.of("-version"), true),
            new ToolDefinition("maven", "Apache Maven", "mvn", List.of("--version"), true),
            new ToolDefinition("git", "Git", "git", List.of("--version"), true),
            new ToolDefinition("node", "Node.js", "node", List.of("--version"), false),
            new ToolDefinition("npm", "npm", "npm", List.of("--version"), false),
            new ToolDefinition("python", "Python", "python", List.of("--version"), false),
            new ToolDefinition("docker", "Docker CLI", "docker", List.of("--version"), false)
    );

    public EnvironmentReport inspect() throws IOException {
        List<CheckResult> results = new ArrayList<>();
        results.add(runtimeResult());
        ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor();
        try {
            List<Future<CheckResult>> futures = TOOLS.stream()
                    .map(tool -> executor.submit(() -> inspectTool(tool)))
                    .toList();
            for (Future<CheckResult> future : futures) {
                try {
                    results.add(future.get());
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new IOException("环境体检已中断", e);
                } catch (ExecutionException e) {
                    Throwable cause = e.getCause();
                    results.add(new CheckResult("unknown", "未知工具", CheckStatus.ERROR,
                            "检测失败", "", errorMessage(cause), false));
                }
            }
        } finally {
            executor.shutdownNow();
        }
        long available = results.stream()
                .filter(result -> result.status() == CheckStatus.AVAILABLE).count();
        long requiredIssues = results.stream()
                .filter(CheckResult::recommended)
                .filter(result -> result.status() != CheckStatus.AVAILABLE).count();
        return new EnvironmentReport(results, (int) available, (int) requiredIssues);
    }

    private CheckResult runtimeResult() {
        String version = System.getProperty("java.runtime.version",
                System.getProperty("java.version", "未知版本"));
        String vendor = System.getProperty("java.vendor", "未知发行方");
        String home = System.getProperty("java.home", "");
        return new CheckResult("runtime", "FxTools Java 运行时", CheckStatus.AVAILABLE,
                version, home, vendor + " · " + System.getProperty("os.arch", ""), false);
    }

    private CheckResult inspectTool(ToolDefinition tool) {
        Optional<Path> executable = findExecutable(tool.command());
        if (executable.isEmpty() && tool.id().equals("python")) {
            executable = findExecutable("py");
        }
        if (executable.isEmpty()) {
            return new CheckResult(tool.id(), tool.name(), CheckStatus.MISSING,
                    "未发现", "", tool.recommended()
                    ? "未在 PATH 中找到，建议配置后重新检测" : "未安装或未加入 PATH",
                    tool.recommended());
        }

        List<String> command = new ArrayList<>();
        command.add(executable.get().toString());
        command.addAll(tool.arguments());
        try {
            CommandExecutionSupport.CommandResult result = CommandExecutionSupport.execute(
                    command, TOOL_TIMEOUT, NATIVE_CHARSET, MAX_OUTPUT_BYTES);
            String firstLine = firstMeaningfulLine(result.output());
            if (result.successful()) {
                String configurationIssue = configurationIssue(tool, executable.get());
                if (configurationIssue != null) {
                    return new CheckResult(tool.id(), tool.name(), CheckStatus.WARNING,
                            firstLine.isBlank() ? "可执行" : firstLine,
                            executable.get().toString(), configurationIssue,
                            tool.recommended());
                }
                return new CheckResult(tool.id(), tool.name(), CheckStatus.AVAILABLE,
                        firstLine.isBlank() ? "可用" : firstLine,
                        executable.get().toString(),
                        result.truncated() ? "输出过长，已截断" : "命令执行正常",
                        tool.recommended());
            }
            return new CheckResult(tool.id(), tool.name(), CheckStatus.ERROR,
                    "命令异常", executable.get().toString(),
                    firstLine.isBlank() ? "退出码 " + result.exitCode() : firstLine,
                    tool.recommended());
        } catch (IOException e) {
            return new CheckResult(tool.id(), tool.name(), CheckStatus.ERROR,
                    "无法执行", executable.get().toString(), errorMessage(e),
                    tool.recommended());
        }
    }

    Optional<Path> findExecutable(String command) {
        if (command == null || command.isBlank()) {
            return Optional.empty();
        }
        Set<String> extensions = executableExtensions();
        String pathValue = System.getenv("PATH");
        if (pathValue == null || pathValue.isBlank()) {
            return Optional.empty();
        }
        for (String directory : pathValue.split(File.pathSeparator)) {
            String cleaned = stripQuotes(directory.strip());
            if (cleaned.isBlank()) {
                continue;
            }
            for (String extension : extensions) {
                try {
                    Path candidate = Path.of(cleaned, command + extension);
                    if (Files.isRegularFile(candidate)) {
                        return Optional.of(candidate.toAbsolutePath().normalize());
                    }
                } catch (InvalidPathException ignored) {
                    // Ignore malformed PATH entries and continue checking the rest.
                }
            }
        }
        return Optional.empty();
    }

    private Set<String> executableExtensions() {
        Set<String> extensions = new LinkedHashSet<>();
        if (WindowsProcessSupport.isWindows()) {
            String pathExt = System.getenv("PATHEXT");
            if (pathExt != null) {
                for (String extension : pathExt.split(";")) {
                    String normalized = extension.strip().toLowerCase(Locale.ROOT);
                    if (!normalized.isBlank()) {
                        extensions.add(normalized.startsWith(".")
                                ? normalized : "." + normalized);
                    }
                }
            }
            extensions.addAll(List.of(".com", ".exe", ".bat", ".cmd"));
        }
        extensions.add("");
        return extensions;
    }

    private String configurationIssue(ToolDefinition tool, Path executable) {
        if (!tool.id().equals("jdk")) {
            return null;
        }
        String javaHome = System.getenv("JAVA_HOME");
        if (javaHome == null || javaHome.isBlank()) {
            return null;
        }
        try {
            Path configuredHome = Path.of(stripQuotes(javaHome.strip()))
                    .toAbsolutePath().normalize();
            Path resolvedExecutable = executable.toAbsolutePath().normalize();
            boolean matches = resolvedExecutable.startsWith(configuredHome);
            return matches ? null : "PATH 中的 javac 与 JAVA_HOME 不一致：" + configuredHome;
        } catch (InvalidPathException e) {
            return "JAVA_HOME 不是有效路径：" + javaHome;
        }
    }

    private static String stripQuotes(String value) {
        return value.length() >= 2 && value.startsWith("\"") && value.endsWith("\"")
                ? value.substring(1, value.length() - 1) : value;
    }

    static String firstMeaningfulLine(String output) {
        if (output == null) {
            return "";
        }
        return output.lines().map(String::strip)
                .filter(line -> !line.isEmpty())
                .findFirst().orElse("");
    }

    private static String errorMessage(Throwable throwable) {
        String message = throwable == null ? null : throwable.getMessage();
        return message == null || message.isBlank()
                ? (throwable == null ? "未知错误" : throwable.getClass().getSimpleName())
                : message;
    }

    public enum CheckStatus {
        AVAILABLE, WARNING, MISSING, ERROR
    }

    public record CheckResult(String id, String name, CheckStatus status, String version,
                              String path, String detail, boolean recommended) {
    }

    public record EnvironmentReport(List<CheckResult> results, int availableCount,
                                    int requiredIssueCount) {
        public EnvironmentReport {
            results = List.copyOf(results);
        }
    }

    private record ToolDefinition(String id, String name, String command,
                                  List<String> arguments, boolean recommended) {
    }
}

File: src/FxTools.Core/Services/DevEnvironmentService.cs
using System.Runtime.InteropServices;

namespace FxTools.Core.Services;

public static class DevEnvironmentService
{
    private static readonly ToolDefinition[] Tools =
    [
        new("jdk", "JDK 编译器", "javac", ["-version"], true),
        new("maven", "Apache Maven", "mvn", ["--version"], true),
        new("git", "Git", "git", ["--version"], true),
        new("node", "Node.js", "node", ["--version"], false),
        new("npm", "npm", "npm", ["--version"], false),
        new("python", "Python", "python", ["--version"], false),
        new("docker", "Docker CLI", "docker", ["--version"], false)
    ];

    public static async Task<EnvironmentReport> InspectAsync(CancellationToken cancellationToken = default)
    {
        Task<CheckResult>[] checks = Tools.Select(
            tool => InspectToolAsync(tool, cancellationToken)).ToArray();
        CheckResult[] toolResults = await Task.WhenAll(checks).ConfigureAwait(false);
        List<CheckResult> results = [RuntimeResult(), .. toolResults];
        return new(
            results,
            results.Count(result => result.Status == CheckStatus.Available),
            results.Count(result => result.Recommended && result.Status != CheckStatus.Available));
    }

    public static string? FindExecutable(string command)
    {
        if (string.IsNullOrWhiteSpace(command))
        {
            return null;
        }
        string? pathValue = Environment.GetEnvironmentVariable("PATH");
        if (string.IsNullOrWhiteSpace(pathValue))
        {
            return null;
        }

        string[] extensions = GetExecutableExtensions();
        foreach (string rawDirectory in pathValue.Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
        {
            string directory = rawDirectory.Trim().Trim('"');
            foreach (string extension in extensions)
            {
                try
                {
                    string candidate = Path.Combine(directory, command + extension);
                    if (File.Exists(candidate))
                    {
                        return Path.GetFullPath(candidate);
                    }
                }
                catch (Exception exception) when (exception is ArgumentException or NotSupportedException or PathTooLongException)
                {
                }
            }
        }
        return null;
    }

    private static async Task<CheckResult> InspectToolAsync(
        ToolDefinition tool,
        CancellationToken cancellationToken)
    {
        string? executable = FindExecutable(tool.Command);
        if (executable is null && tool.Id == "python")
        {
            executable = FindExecutable("py");
        }
        if (executable is null)
        {
            return new(tool.Id, tool.Name, CheckStatus.Missing, "未发现", string.Empty,
                tool.Recommended ? "未在 PATH 中找到，建议配置后重新检测" : "未安装或未加入 PATH",
                tool.Recommended);
        }

        try
        {
            CommandResult result = await BoundedCommandRunner.RunAsync(
                executable, tool.Arguments, TimeSpan.FromSeconds(5), 256 * 1024,
                cancellationToken).ConfigureAwait(false);
            string firstLine = FirstMeaningfulLine(result.Output);
            if (!result.Successful)
            {
                return new(tool.Id, tool.Name, CheckStatus.Error, "命令异常", executable,
                    string.IsNullOrEmpty(firstLine) ? $"退出码 {result.ExitCode}" : firstLine,
                    tool.Recommended);
            }

            string? issue = ConfigurationIssue(tool, executable);
            return new(tool.Id, tool.Name,
                issue is null ? CheckStatus.Available : CheckStatus.Warning,
                string.IsNullOrEmpty(firstLine) ? "可用" : firstLine,
                executable,
                issue ?? (result.Truncated ? "输出过长，已截断" : "命令执行正常"),
                tool.Recommended);
        }
        catch (OperationCanceledException) { throw; }
        catch (Exception exception) when (exception is IOException or TimeoutException)
        {
            return new(tool.Id, tool.Name, CheckStatus.Error, "无法执行", executable,
                exception.Message, tool.Recommended);
        }
    }

    private static CheckResult RuntimeResult() => new(
        "runtime",
        "FxTools .NET 运行时",
        CheckStatus.Available,
        RuntimeInformation.FrameworkDescription,
        AppContext.BaseDirectory,
        $"{RuntimeInformation.OSDescription} · {RuntimeInformation.ProcessArchitecture}",
        false);

    private static string? ConfigurationIssue(ToolDefinition tool, string executable)
    {
        if (tool.Id != "jdk")
        {
            return null;
        }
        string? javaHome = Environment.GetEnvironmentVariable("JAVA_HOME")?.Trim().Trim('"');
        if (string.IsNullOrWhiteSpace(javaHome))
        {
            return null;
        }
        try
        {
            return Path.GetFullPath(executable).StartsWith(
                Path.GetFullPath(javaHome), StringComparison.OrdinalIgnoreCase)
                ? null
                : $"PATH 中的 javac 与 JAVA_HOME 不一致: {javaHome}";
        }
        catch (Exception exception) when (exception is ArgumentException or NotSupportedException or PathTooLongException)
        {
            return $"JAVA_HOME 不是有效路径: {javaHome}";
        }
    }

    private static string[] GetExecutableExtensions()
    {
        if (!OperatingSystem.IsWindows())
        {
            return [string.Empty];
        }
        string? pathExt = Environment.GetEnvironmentVariable("PATHEXT");
        IEnumerable<string> configured = (pathExt ?? ".COM;.EXE;.BAT;.CMD")
            .Split(';', StringSplitOptions.RemoveEmptyEntries)
            .Select(extension => extension.Trim().ToLowerInvariant())
            .Select(extension => extension.StartsWith('.') ? extension : $".{extension}");
        return configured.Concat([".com", ".exe", ".bat", ".cmd", string.Empty])
            .Distinct(StringComparer.OrdinalIgnoreCase).ToArray();
    }

    public static string FirstMeaningfulLine(string? output) => output?
        .Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
        .FirstOrDefault() ?? string.Empty;

    private sealed record ToolDefinition(
        string Id,
        string Name,
        string Command,
        IReadOnlyList<string> Arguments,
        bool Recommended);
}

public sealed record EnvironmentReport(
    IReadOnlyList<CheckResult> Results,
    int AvailableCount,
    int RequiredIssueCount);

public sealed record CheckResult(
    string Id,
    string Name,
    CheckStatus Status,
    string Version,
    string Path,
    string Detail,
    bool Recommended);

public enum CheckStatus
{
    Available,
    Warning,
    Missing,
    Error
}
